interfaces/seccomp/backend.go

```diff
@@ -209,3 +227,56 @@ func (b *Backend) SandboxFeatures() []string {
 	tags = append(tags, "bpf-argument-filtering")
 	return tags
 }
+
+// Determine if the system requires the use of socketcall(). Factors:
+// - if the kernel architecture is amd64, armhf or arm64, do not require
+//   socketcall (unused on these architectures)
+// - if the kernrel architecture is i386
+//   - if the kernel is < 4.3, force the use of socketcall()
+//   - for backwards compatibility, if the system is Ubuntu 14.04 or lower,
+//     force use of socketcall()
+//   - for backwards compatibility, if the base snap is unspecified, "core" or
+//     "core16", then force use of socketcall()
+//   - otherwise (ie, if new enough kernel, not 14.04, and a non-16 base
+//     snap), don't force use of socketcall()
+// - if the kernel architecture is not any of the above, force the use of
+//   socketcall() (TODO: verify)
+func RequiresSocketcall(baseSnap string) bool {
+	var needed bool
+
+	switch ubuntuKernelArchitecture() {
+	case "i386":
+		needed = false
+		if cmp, _ := strutil.VersionCompare(kernelVersion(), "4.3"); cmp < 0 {
+			// On kernels < 4.3, always require socketcall(). See
+			// 'man 2 socketcall'
+			needed = true
+		} else if releaseInfoId == "ubuntu" {
+			// For now, on 14.04, always require socketcall()
+			if cmp, _ := strutil.VersionCompare(releaseInfoVersionId, "14.04"); cmp <= 0 {
+				needed = true
+			}
+		} else {
+			// Detect when the base snap requires the use of
+			// socketcall().
+			//
+			// TODO: eventually try to auto-detect this. For now,
+			// err on the side of security and only require it for
+			// base snaps where we know we want it added.
+			// Technically, core16's glibc is new enough, but it
+			// always had socketcall in the template, so ensure
+			// backwards compatibility.
+			if baseSnap == "" || baseSnap == "core" || baseSnap == "core16" {
+				needed = true
+			}
+		}
+	case "powerpc", "ppc64el", "s390x":
+		// TBD
+		needed = true
```

Comment:
FYI, I've thus far been unable to test these architectures, so I checked unistd.h on these 3 architectures, and unlike i386, they have the individual socket calls defined. We could probably just default to `needed = false`. Another option is on these, add the above baseSnap check to ensure existing installs aren't affected, then see how it goes with core18.

Reply:
Ah, reading the glibc sources, we see that ./sysdeps/unix/sysv/linux/kernel-features.h says __ASSUME_SOCKETCALL needs to be defined on the architectures that support socketcall(). Looking for that, I see that:
 * sysdeps/unix/sysv/linux/powerpc/kernel-features.h: powerpc has individual socket syscalls as of 2.6.37
 * ./sysdeps/unix/sysv/linux/s390/kernel-features.h: s390 has individual socket syscalls as of 4.3
 * ./sysdeps/unix/sysv/linux/m68k/kernel-features.h: m68k has individual socket syscalls as of 4.3
 * ./sysdeps/unix/sysv/linux/sh/kernel-features.h: sh has individual socket syscalls as of 2.6.37
 * ./sysdeps/unix/sysv/linux/microblaze/kernel-features.h: microblaze has all individual socket syscalls as of 2.6.37

I'll adjust the PR accordingly.

Reply:
Done.